src/main.cpp

```diff
@@ -866,11 +910,50 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
 {
     // Basic checks that don't depend on any context
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    /**
+     * Previously:
+     * 1. The consensus rule below was:
+     *        if (tx.nVersion < SPROUT_MIN_TX_VERSION) { ... }
+     *    which checked if tx.nVersion fell within the range:
+     *        INT32_MIN <= tx.nVersion < SPROUT_MIN_TX_VERSION
+     * 2. The parser allowed tx.nVersion to be negative
+     *
+     * Now:
+     * 1. The consensus rule checks to see if tx.Version falls within the range:
+     *        0 <= tx.nVersion < SPROUT_MIN_TX_VERSION
+     * 2. The previous consensus rule checked for negative values within the range:
+     *        INT32_MIN <= tx.nVersion < 0
+     *    This is unnecessary for Overwinter transactions since the parser now
+     *    interprets the sign bit as fOverwintered, so tx.nVersion is always >=0,
+     *    and when Overwinter is not active ContextualCheckTransaction rejects
+     *    transactions with fOverwintered set.  When Overwinter has activated and
+     *    fOverwintered is set, this function CheckTransactionWithoutProofVerification
+     *    checks to make sure that tx.nVersion avoids the following ranges:
+     *        0 <= tx.nVersion < OVERWINTER_MIN_TX_VERSION
+     *        OVERWINTER_MAX_TX_VERSION < tx.nVersion <= INT32_MAX
+     */
+    if (!tx.fOverwintered && tx.nVersion < SPROUT_MIN_TX_VERSION) {
         return state.DoS(100, error("CheckTransaction(): version too low"),
                          REJECT_INVALID, "bad-txns-version-too-low");
     }
+    else if (tx.fOverwintered) {
+        if (tx.nVersion < OVERWINTER_MIN_TX_VERSION) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too low"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-low");
+        }
+        if (tx.nVersion > OVERWINTER_MAX_TX_VERSION ) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too high for client"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-high");
+        }
+        if (tx.nVersionGroupId != OVERWINTER_VERSION_GROUP_ID) {
+            return state.DoS(100, error("CheckTransaction(): unknown tx version group id"),
+                    REJECT_INVALID, "bad-tx-version-group-id");
+        }
+        if (tx.nExpiryHeight > TX_EXPIRY_HEIGHT_THRESHOLD) {
```

Comment:
``LOCKTIME_THRESHOLD = 500000000`` is exclusive, i.e. a locktime is a Unix timestamp if it is *less than* 500000000.
But ``TX_EXPIRY_HEIGHT_THRESHOLD = 500000000`` as implemented here (and as specified in [ZIP 203](https://github.com/zcash/zips/pull/131/files)) is inclusive, i.e. the value 500000000 is accepted and treated as a block height rather than a Unix timestamp. This is inconsistent (and would result in a very ugly wart if we decided to extend ``nExpiryHeight`` to allow the timestamp meaning in future).

Reply:
Good spot, the bitcoin wiki says here https://en.bitcoin.it/wiki/Protocol_documentation#tx
```
< 500000000 | Block number at which this transaction is unlocked
>= 500000000 | UNIX timestamp at which this transaction is unlocked
```
Will update the code.  @arcalinea can you please do the same with the ZIP?  Thanks.